invokeai/backend/model_manager/download/queue.py

```diff
@@ -0,0 +1,689 @@
+# Copyright (c) 2023, Lincoln D. Stein
+"""Implementation of multithreaded download queue for invokeai."""
+
+import os
+import re
+import shutil
+import threading
+import time
+import traceback
+from pathlib import Path
+from queue import PriorityQueue
+from typing import Dict, List, Optional, Set, Tuple, Union
+
+import requests
+from huggingface_hub import HfApi, hf_hub_url
+from pydantic import Field, ValidationError, parse_obj_as, validator
+from pydantic.networks import AnyHttpUrl
+from requests import HTTPError
+
+from invokeai.app.services.config import InvokeAIAppConfig
+from invokeai.backend.util import InvokeAILogger, Logger
+
+from ..storage import DuplicateModelException
+from . import HTTP_RE, REPO_ID_RE
+from .base import (
+    DownloadEventHandler,
+    DownloadJobBase,
+    DownloadJobStatus,
+    DownloadQueueBase,
+    ModelSourceMetadata,
+    UnknownJobIDException,
+)
+
+# Maximum number of bytes to download during each call to requests.iter_content()
+DOWNLOAD_CHUNK_SIZE = 100000
+
+# marker that the queue is done and that thread should exit
+STOP_JOB = DownloadJobBase(id=-99, priority=-99, source="dummy", destination="/")
+
+# endpoint for civitai get-model API
+CIVITAI_MODEL_DOWNLOAD = r"https://civitai.com/api/download/models/(\d+)"
+CIVITAI_MODEL_PAGE = "https://civitai.com/models/"
+CIVITAI_MODEL_PAGE_WITH_VERSION = r"https://civitai.com/models/(\d+)\?modelVersionId=(\d+)"
+CIVITAI_MODELS_ENDPOINT = "https://civitai.com/api/v1/models/"
+CIVITAI_VERSIONS_ENDPOINT = "https://civitai.com/api/v1/model-versions/"
+
+
+class DownloadJobURL(DownloadJobBase):
+    """Job declaration for downloading individual URLs."""
+
+    source: AnyHttpUrl = Field(description="URL to download")
+
+
+class DownloadJobRepoID(DownloadJobBase):
+    """Download repo ids."""
+
+    variant: Optional[str] = Field(description="Variant, such as 'fp16', to download")
+    source: str = Field(description="URL or repo_id to download")
+
+    @validator("source")
+    @classmethod
+    def _validate_source(cls, v: str) -> str:
+        if not re.match(REPO_ID_RE, v):
+            raise ValidationError(f"{v} invalid repo_id", cls)
+        return v
+
+
+class DownloadJobPath(DownloadJobBase):
+    """Handle file paths."""
+
+    source: Union[str, Path] = Field(description="Path to a file or directory to install")
+
+
+class DownloadQueue(DownloadQueueBase):
```

Comment:
Since you did your review, I have split `queue.py` into two parts. One does basic queued downloading of URLs and local paths, and is called `queue.py`. The other handles the logic of parallel fetch from HuggingFace repo_ids and acquisition of model metadata, and is called `model_queue.py`. When you have a chance check and see if this is easier to follow. I could certainly split up the queue functionality further, but take a look first.